coderd/authzquery/authz.go

```diff
@@ -0,0 +1,314 @@
+			slog.F("input", internalError.Input()),
+			slog.Error(err),
+		)
+	}
+	return NotAuthorizedError{
+		Err: err,
+	}
+}
+
+func authorizedInsert[ArgumentType any,
+	Insert func(ctx context.Context, arg ArgumentType) error](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	action rbac.Action,
+	object rbac.Objecter,
+	insertFunc Insert) Insert {
+	return func(ctx context.Context, arg ArgumentType) error {
+		_, err := authorizedInsertWithReturn(logger, authorizer, action, object, func(ctx context.Context, arg ArgumentType) (rbac.Objecter, error) {
+			return rbac.Object{}, insertFunc(ctx, arg)
+		})(ctx, arg)
+		return err
+	}
+}
+
+func authorizedInsertWithReturn[ObjectType any, ArgumentType any,
+	Insert func(ctx context.Context, arg ArgumentType) (ObjectType, error)](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	action rbac.Action,
+	object rbac.Objecter,
+	insertFunc Insert) Insert {
+	return func(ctx context.Context, arg ArgumentType) (empty ObjectType, err error) {
+		// Fetch the rbac subject
+		act, ok := ActorFromContext(ctx)
+		if !ok {
+			return empty, NoActorError
+		}
+
+		// Authorize the action
+		err = authorizer.Authorize(ctx, act, action, object.RBACObject())
+		if err != nil {
+			return empty, LogNotAuthorizedError(ctx, logger, err)
+		}
+
+		// Insert the database object
+		return insertFunc(ctx, arg)
+	}
+}
+
+func authorizedDelete[ObjectType rbac.Objecter, ArgumentType any,
+	Fetch func(ctx context.Context, arg ArgumentType) (ObjectType, error),
+	Delete func(ctx context.Context, arg ArgumentType) error](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	fetchFunc Fetch,
+	deleteFunc Delete) Delete {
+	return authorizedFetchAndExec(logger, authorizer,
+		rbac.ActionDelete, fetchFunc, deleteFunc)
+}
+
+func authorizedUpdateWithReturn[ObjectType rbac.Objecter,
+	ArgumentType any,
+	Fetch func(ctx context.Context, arg ArgumentType) (ObjectType, error),
+	UpdateQuery func(ctx context.Context, arg ArgumentType) (ObjectType, error)](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	fetchFunc Fetch,
+	updateQuery UpdateQuery) UpdateQuery {
+	return authorizedFetchAndQuery(logger, authorizer, rbac.ActionUpdate, fetchFunc, updateQuery)
+}
+
+func authorizedUpdate[ObjectType rbac.Objecter,
+	ArgumentType any,
+	Fetch func(ctx context.Context, arg ArgumentType) (ObjectType, error),
+	Exec func(ctx context.Context, arg ArgumentType) error](
+	// Arguments
+	logger slog.Logger,
+	authorizer rbac.Authorizer,
+	fetchFunc Fetch,
+	updateExec Exec) Exec {
+	return authorizedFetchAndExec(logger, authorizer, rbac.ActionUpdate, fetchFunc, updateExec)
+}
+
+// authorizedFetchAndExecWithConverter uses authorizedFetchAndQueryWithConverter but
+// only cares about the error return type. SQL execs only return an error.
+// See authorizedFetchAndQueryWithConverter for more details.
+func authorizedFetchAndExec[ObjectType rbac.Objecter,
```

Comment:
having everything start with `authorized` feels tedious and doesn't match typical Golang style of omitting contextual prefixes